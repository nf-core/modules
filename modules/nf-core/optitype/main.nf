process OPTITYPE {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/mulled-v2-f3579993d544920cc7c09e8ddf77569e95ea8950:d3ee2331cb4615915b1eda5d23f231da6b0741d7-0' :
        'biocontainers/mulled-v2-f3579993d544920cc7c09e8ddf77569e95ea8950:d3ee2331cb4615915b1eda5d23f231da6b0741d7-0' }"

    input:
    tuple val(meta), path(bam), path(bai)

    output:
    tuple val(meta), path("${prefix}/*.tsv"), emit: hla_type
    tuple val(meta), path("${prefix}/*.pdf"), emit: coverage_plot
    path "versions.yml"                     , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args  = task.ext.args   ?: ''
    def args2 = task.ext.args2  ?: ''
    prefix    = task.ext.prefix ?: "${meta.id}"

    """
    # Create a config for OptiType on a per sample basis with task.ext.args2

    #Doing it old school now
    echo "[mapping]" > config.ini
    echo "razers3=razers3" >> config.ini
    echo "threads=$task.cpus" >> config.ini
    echo "[ilp]" >> config.ini
    echo "$args2" >> config.ini
    echo "threads=1" >> config.ini
    echo "[behavior]" >> config.ini
    echo "deletebam=true" >> config.ini
    echo "unpaired_weight=0" >> config.ini
    echo "use_discordant=false" >> config.ini

    # Run the actual OptiType typing with args
    OptiTypePipeline.py -i ${bam} -c config.ini --${meta.seq_type} $args --prefix $prefix --outdir $prefix

    #Couldn't find a nicer way of doing this
    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        optitype: \$(cat \$(which OptiTypePipeline.py) | grep -e "Version:" | sed -e "s/Version: //g")
    END_VERSIONS
    """
}
