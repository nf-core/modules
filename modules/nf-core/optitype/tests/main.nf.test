nextflow_process {

    name "Test Process OPTITYPE"
    script "../main.nf"
    process "OPTITYPE"

    tag "modules"
    tag "modules_nfcore"
    tag "optitype"

    test("hla - bam") {
        config "./nextflow.config"
        when {
            process {
                """
                input[0] = [ [ id:'test', seq_type:'dna' ], // meta map
                       file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/example_hla_pe.sorted.bam', checkIfExists: true),
                       file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/example_hla_pe.sorted.bam.bai', checkIfExists: true)
                     ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(file(process.out.hla_type[0][1])
                    .readLines()[1]
                    .toString()
                    // the calls seem not to be stable for locus A for this input data, therefore skip the results for HLA locus A
                    .split('\t')[3..7])
                    .match("bam_hla_type") },
                { assert snapshot(file(process.out.coverage_plot[0][1]).name).match("bam_coverage_plot") },
                { assert snapshot(process.out.versions).match("bam_versions") }
            )
        }
    }

    test("hla - bam - cbc") {
        config "./optitype_cbc.config"
        when {
            process {
                """
                input[0] = [ [ id:'test', seq_type:'dna' ], // meta map
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/example_hla_pe.sorted.bam', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/example_hla_pe.sorted.bam.bai', checkIfExists: true)
                     ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(file(process.out.hla_type[0][1])
                    .readLines()[1]
                    .toString()
                    // the calls seem not to be stable for locus A for this input data, therefore skip the results for HLA locus A
                    .split('\t')[3..7])
                    .match("bam_cbc_hla_type") },
                { assert snapshot(file(process.out.coverage_plot[0][1]).name).match("bam_cbc_coverage_plot") },
                { assert snapshot(process.out.versions).match("bam_cbc_versions") }
            )
        }
    }

}
