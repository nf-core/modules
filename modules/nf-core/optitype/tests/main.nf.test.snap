{
    "bam_coverage_plot": {
        "content": [
            "test_coverage_plot.pdf"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-06T13:47:07.297380921"
    },
    "bam_cbc_versions": {
        "content": [
            [
                "versions.yml:md5,b605daa15a7260619625e662566bffa1"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-06T13:50:18.894709282"
    },
    "bam_cbc_coverage_plot": {
        "content": [
            "test_coverage_plot.pdf"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-06T13:50:18.890519853"
    },
    "bam_hla_type": {
        "content": [
            [
                "B*08:01",
                "B*57:01",
                "C*06:02",
                "C*07:01",
                "1439.0"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-06T13:47:07.279221023"
    },
    "bam_versions": {
        "content": [
            [
                "versions.yml:md5,b605daa15a7260619625e662566bffa1"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-06T13:47:07.301900201"
    },
    "bam_cbc_hla_type": {
        "content": [
            [
                "B*08:01",
                "B*57:01",
                "C*06:02",
                "C*07:01",
                "1439.0"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-06T13:50:18.885995054"
    }
}